wgsl/index.bs

```diff
@@ -3346,29 +3346,94 @@ and vice versa,
+<dfn noexport>valid pointer</dfn> is a pointer that corresponds to a [=valid reference=],
+and an <dfn noexport>invalid pointer</dfn> is a pointer that corresponds to an [=invalid memory reference=].
+
 ### Originating Variable ### {#originating-variable-section}
 
-In WGSL a reference value always corresponds to the memory view
-for some or all of the memory locations for some variable.
+A [=valid reference=] corresponds to the memory view for some or all of the memory locations for some variable.
 This defines the <dfn noexport>originating variable</dfn> for the reference value.
 
-A pointer value always corresponds to a reference value, and so the originating variable
-of a pointer is the same as the originating variable of the corresponding reference.
+The [=originating variable=] of a [=valid pointer=] is the same as the originating variable of the corresponding valid reference.
 
 Note: The originating variable is a dynamic concept.
 The originating variable for a formal parameter of a function depends on the
 [=call site|call sites=] for the function.
 Different call sites may supply pointers into different originating variables.
 
-### Invalid Memory Reference ### {#invalid-mem-reference}
+<div class=note>
+<span class=marker>Note:</span> A reference can correspond to memory locations inside a variable, and still be invalid.
+This can occur when an index is too large for the type being indexed, but the referenced locations
+would be inside a subsequent sibling data member.
+
+In the following example, the reference `the_particle.position[i]` is valid if and only if `i` is 0 or 1.
+When `i` is 2, the reference will be an [=invalid memory reference=], but would otherwise correspond
+the memory locations for `the_particle.color_index`.
+<div class='example wgsl' heading='Invalid memory reference still inside a variable'>
+<span id="example-invalid-ref"></span>
+  <xmp highlight='rust'>
+    struct Particle {
+       position: vec2f,
+       velocity: vec2f,
+       color_index: i32,
+    }
 
-An operation that accesses one or more [=memory locations=] outside those
-associated with the operation's [=memory view=] is an
-<dfn noexport>out of bounds access</dfn>.
-Any reference or pointer that, when accessed, would produce an out of bounds
-access is an <dfn noexport>invalid memory reference</dfn>.
+    @group(0) @binding(0)
+    var<storage,read_write> the_particle: Particle;
 
-[=Load Rule|Loads=] from an invalid reference return one of:
+    fn particle_velocity_component(p: Particle, i: i32) -> f32 {
+      return the_particle.velocity[i]; // A valid reference when i is 0 or 1.
+    }
+  </xmp>
+</div>
+</div>
+
+### Out of Bounds Access ### {#out-of-bounds-access-sec}
+
+An operation that [=memory access|accesses=] an
+[=invalid memory reference=] is an <dfn noexport>out of bounds access</dfn>.
+It is a [=dynamic error=] to perform an [=out of bounds access=].
+
+An out of bounds access, if performed as written, would typically:
```

Comment:
Maybe make this sound less prescriptive? Maybe even just make this a note
```suggestion
Note: The behavior of an out of bounds access may include:
```

Reply:
You go on to describe trapping as a possible result which doesn't mesh well with this text.

Reply:
Yea, need something better here. Should say these things *don't* actually happen.